### DISCUSSION TITLE
EventSets handling code severely outdated

### DISCUSSION THREAD

**Discussion by UnhappyWithMyUsername** [FakeFishGames/Barotrauma#14014] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [x] My issue happened while using mods.

### What happened?

The way EventSets are handled is severely outdated and causes event mods to never be fully compatible due to the way the sets are loaded.

The game relies on seemingly extremely old code that loads a set, then checks if that set is additive. If it is in fact additive, it will lock that set in and load another set that isn't configured as additive, then proceeds with generating the level.
If the game managed to load a set that isn't additive in the first attempt, it will instantly proceed to generate the level, not even looking for an additive set that might've been configured to always load, resulting in a level with only the main EventSet running.

This of course breaks event mods that for example rely on a Set that should always run alongside the main set, but due to there not being a guarantee for additive sets to load in, it is not possible to achieve that.
Only allowing for one additive set to be run alongside the main event set is also causing issues if multiple mods introduce events for the same locations as only one of them will have the chance to load in, potentially breaking the mod.
Additive should mean, the event set will always load alongside the main set as long as its conditions are met.

The code in question:
https://github.com/FakeFishGames/Barotrauma/blob/a35dbdc87cc4c07606f79921e10aeda1d4d60f67/Barotrauma/BarotraumaShared/SharedSource/Events/EventManager.cs#L182-L205

### Reproduction steps

1. Create a mod with multiple event sets and configure them with additive="true"
2. Configure ScriptedEvents inside your sets and configure them to always load.
3. Load into the campaign and check the loaded events with "debugai". You may only see the events of one of your additive sets at all times and perhaps not even see a single one of them.

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Happens in both single player and multiplayer

### Version

v1.4.6.0 (Blood in the Water Update, hotfix 2)

### Which operating system did you encounter this bug on?

Windows

**Comment by Regalis11** [comment]
Thank you for the report!

To me this seems more like a feature request rather than a bug report. There is currently no way to guarantee an additive set is always selected in addition to the "normal set". I do agree that the way it works makes it difficult to use this functionality in mods, so I think we should look into changing it.

**Reply by UnhappyWithMyUsername** [reply]
Yes I agree this technically is more of a feature request, tho I gotta be honest here, I fear it will be lost and forgotten as a feature request.

And yes it would be greatly appreciated if this could be addressed. This is basically the number one reason for one of the most common bug reports on DynamicEuropa about events missing and there's currently just no way for us to make sure everything is compatible other than telling people to not use other mods or making a compatibility patches for every mod that adds custom events.

**Comment by itchyOwl** [comment]
I don't quite understand your argument about "extremely old" and "severely outdated" code? The support for additive sets was introduced in July 2021 (FakeFishGames/Barotrauma-development@588af7d0), and it's been revisited a few times after that. Was there something changed in the game in your opinion that makes this code outdated?

**Reply by UnhappyWithMyUsername** [reply]
That's just how it feels. It doesn't properly support adding events alongside the vanilla events without messing something up.
Why would anyone bother making events when they can't make sure they are working as intended unless they override everything and make their mod basically incompatible with others?

When I worked on events, I expected to just put my events into a new event set, configure it for the locations the events should happen at and set it to additive="true" so it **always** adds itself alongside other sets that may be running on these locations.
But that's not the case. Your set may trigger if you're lucky or it may not.

If you have multiple mods that introduce new event sets, they all fight each other over that one additional set spot that may or may not even get loaded. IMO they should all just load alongside each other when they use additive="true" since as far as I know additive stands for an addition, not a replacement from a logical standpoint.

So in short: This current code doesn't allow properly adding new events that actually trigger when they should unless you don't define them as additive and instead copy and override the entire original eventset from the game, making your mod less compatible.

**Reply by Regalis11** [reply]
I agree that the way it works makes it difficult to use in mods, but I don't see how the code is "outdated", nor why it'd be a problem that some code is extremely old. The intended behavior was that the game can randomly choose one additive set in addition to the "normal set", not that all additive sets always run, and that still works as intended despite the code being old - it's just that the intended behavior is not very user friendly.

But yeah I'm getting too hung up on the choices of words here, I get what you mean and agree there's an issue here we should address!

**Comment by Regalis11** [comment]
Addressed in https://github.com/FakeFishGames/Barotrauma-development/commit/76cbe828d6e38c1c26d6909bde93fe2cd4912c0f

> Made it easier for modders to add event sets in their mods and guarantee they get selected. We have support for "additive sets", which can get selected on top of a normal non-additive set, but it's entirely random: there's no way to guarantee your modded sets get chosen. I added a new `SelectAlways` setting to event sets, which can be used to guarantee the set gets selected.

**Reply by UnhappyWithMyUsername** [reply]
Oh wow that was fast. Absolute legend! Thank you so much for this!

**Comment by Jasontti** [comment]
Tested fix for this and found no issues.
Closing discussion as resolved.
